File: wingfoil/examples/zmq/direct/zmq_sub.rs
// ZMQ subscriber — direct mode (no service discovery).
//
// Connects to the publisher address directly. Start `zmq_direct_pub` first.
// Cross-language compatible — Python pub works too.
//
// Run publisher and subscriber in separate terminals:
//
//   RUST_LOG=info cargo run --example zmq_direct_pub --features zmq
//   RUST_LOG=info cargo run --example zmq_direct_sub --features zmq

use log::Level::Info;
use wingfoil::adapters::zmq::zmq_sub;
use wingfoil::*;

fn main() -> anyhow::Result<()> {
    env_logger::init();

    let (data, _status) = zmq_sub::<Vec<u8>>("tcp://127.0.0.1:7779")?;
    data.map(|burst| {
        burst
            .into_iter()
            .map(|b| String::from_utf8_lossy(&b).into_owned())
            .collect::<Vec<_>>()
    })
    .logged("sub", Info)
    .run(RunMode::RealTime, RunFor::Forever)?;

    Ok(())
}

File: legacy/wingfoil-python/examples/zmq/direct/zmq_sub.py
#!/usr/bin/env python3
"""ZMQ subscriber — direct mode.

Connects to the publisher at ADDRESS. Start zmq_pub.py first.
Cross-language compatible — the Rust publisher works too.

Run publisher and subscriber in separate terminals:

    python examples/zmq/direct/zmq_pub.py
    python examples/zmq/direct/zmq_sub.py
"""

import sys
import wingfoil as wf

ADDRESS = "tcp://127.0.0.1:7779"
print(f"Connecting to {ADDRESS} ... (Ctrl-C to exit)")

data, status = wf.zmq_sub(ADDRESS)

data_node = data.inspect(lambda msgs: [
    print(f"received: {m.decode()}", flush=True) for m in msgs
])

status_node = status.inspect(lambda s: print(f"status: {s}", flush=True))

try:
    wf.Graph([data_node, status_node]).run(realtime=True)
except KeyboardInterrupt:
    print("\nExiting.")
    sys.exit(0)
